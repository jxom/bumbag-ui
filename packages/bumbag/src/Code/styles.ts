import { css, cssClass } from '../styled';
import { borderRadius, font, palette, theme } from '../utils';

export const Code = (styleProps) => cssClass`
  background-color: ${
    styleProps.palette === 'default'
      ? palette('white700', { dark: 'black300' })(styleProps)
      : palette(`${styleProps.palette}Tint`, { dark: `${styleProps.palette}Shade` })(styleProps)
  };
  border-radius: ${borderRadius('1')(styleProps)};
  color: ${
    styleProps.palette === 'default'
      ? palette('text')(styleProps)
      : palette(`${styleProps.palette}700`, { dark: `${styleProps.palette}200` })(styleProps)
  };
  font-family: ${font('mono')(styleProps)};
  padding: 0.1rem 0.2rem;
  overflow: auto;

  & {
    ${
      styleProps.isBlock &&
      css`
        padding: 1rem;
        ${theme(styleProps.themeKey, `styles.block`)(styleProps)};
      `
    }
  };

  & {
    ${theme(styleProps.themeKey, `styles.base`)(styleProps)};
  };
`;
