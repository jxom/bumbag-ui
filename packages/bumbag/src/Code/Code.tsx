import { Box as ReakitBox } from 'reakit';

import { Palette } from '../types';
import { useClassName, createComponent, createElement, createHook } from '../utils';
import { Box, BoxProps } from '../Box';

import * as styles from './styles';

export type LocalCodeProps = {
  /** Indicates if the code should be placed in a block. */
  isBlock?: boolean;
  /** Color of the code (inline) block. */
  palette?: Palette;
};
export type CodeProps = BoxProps & LocalCodeProps;

const useProps = createHook<CodeProps>(
  (props, { themeKey, themeKeyOverride }) => {
    const boxProps = Box.useProps(props);

    const className = useClassName({
      style: styles.Code,
      styleProps: props,
      themeKey,
      themeKeyOverride,
      prevClassName: boxProps.className,
    });

    return { ...boxProps, className };
  },
  {
    defaultProps: {
      palette: 'default',
    },
    themeKey: 'Code',
  }
);

export const Code = createComponent<CodeProps>(
  (props) => {
    let use = props.use;
    if (!use) {
      use = props.isBlock ? 'pre' : 'code';
    }

    const codeProps = useProps(props);

    return createElement({ children: props.children, component: ReakitBox, use, htmlProps: codeProps });
  },
  {
    attach: {
      useProps,
      displayName: 'Code',
    },
    themeKey: 'Code',
  }
);
