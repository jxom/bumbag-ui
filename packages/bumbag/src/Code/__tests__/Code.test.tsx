import * as React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { Code } from '../Code';
import render from '../../utils/_tests/render';

describe('props', () => {
  it('should render correctly', () => {
    const { container } = render(<Code>Hello world</Code>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with CSS props', () => {
    const { container } = render(<Code color="primary">Hello world</Code>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly as a block', () => {
    const { container } = render(<Code isBlock>Hello world</Code>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with palette', () => {
    const { container } = render(<Code palette="primary">Hello world</Code>);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('composition', () => {
  describe('as', () => {
    it('should render correctly', () => {
      const { container } = render(<Code use="p">Hello world</Code>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('hook', () => {
    it('should return with Code props', () => {
      const { result } = renderHook(() => Code.useProps());
      expect(result.current).toMatchSnapshot();
    });
  });

  describe('render props', () => {
    it('should render correctly', () => {
      const { container } = render(<Code>{(CodeProps) => <div {...CodeProps}>Hello world</div>}</Code>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});

describe('theming', () => {
  it('Code.base should render correctly', () => {
    const { container } = render(<Code>hello world</Code>, {
      theme: { Code: { styles: { base: { backgroundColor: 'red' } } } },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultProps', () => {
  it('should render correctly for className', () => {
    const { container } = render(<Code>hello world</Code>, {
      theme: { Code: { defaultProps: { className: 'test' } } },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly for palette', () => {
    const { container } = render(<Code>hello world</Code>, {
      theme: { Code: { defaultProps: { palette: 'primary' } } },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
