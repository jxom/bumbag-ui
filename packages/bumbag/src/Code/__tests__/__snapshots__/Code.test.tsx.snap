// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composition as should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  background-color: var(--bb-palette-white700,#000000);
  border-radius: 2px;
  color: var(--bb-palette-text,#000000);
  font-family: 'SF Mono','Segoe UI Mono','Roboto Mono',Menlo,Courier,monospace;
  padding: 0.1rem 0.2rem;
  overflow: auto;
}

<p
  class="emotion-0 emotion-1 emotion-2 bb-Box bb-Code"
>
  Hello world
</p>
`;

exports[`composition hook should return with Code props 1`] = `
Object {
  "className": "css-1wdukht css-63it0p css-jwhr8m bb-Box bb-Code",
}
`;

exports[`composition render props should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  background-color: var(--bb-palette-white700,#000000);
  border-radius: 2px;
  color: var(--bb-palette-text,#000000);
  font-family: 'SF Mono','Segoe UI Mono','Roboto Mono',Menlo,Courier,monospace;
  padding: 0.1rem 0.2rem;
  overflow: auto;
}

<div
  class="emotion-0 emotion-1 emotion-2 bb-Box bb-Code"
>
  Hello world
</div>
`;

exports[`defaultProps should render correctly for className 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  background-color: var(--bb-palette-white700,#000000);
  border-radius: 2px;
  color: var(--bb-palette-text,#000000);
  font-family: 'SF Mono','Segoe UI Mono','Roboto Mono',Menlo,Courier,monospace;
  padding: 0.1rem 0.2rem;
  overflow: auto;
}

<code
  class="emotion-0 emotion-1 emotion-2 test bb-Box bb-Code"
>
  hello world
</code>
`;

exports[`defaultProps should render correctly for palette 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  background-color: var(--bb-palette-primaryTint,#000000);
  border-radius: 2px;
  color: var(--bb-palette-primary700,#000000);
  font-family: 'SF Mono','Segoe UI Mono','Roboto Mono',Menlo,Courier,monospace;
  padding: 0.1rem 0.2rem;
  overflow: auto;
}

<code
  class="emotion-0 emotion-1 emotion-2 bb-Box bb-Code"
>
  hello world
</code>
`;

exports[`props should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  background-color: var(--bb-palette-white700,#000000);
  border-radius: 2px;
  color: var(--bb-palette-text,#000000);
  font-family: 'SF Mono','Segoe UI Mono','Roboto Mono',Menlo,Courier,monospace;
  padding: 0.1rem 0.2rem;
  overflow: auto;
}

<code
  class="emotion-0 emotion-1 emotion-2 bb-Box bb-Code"
>
  Hello world
</code>
`;

exports[`props should render correctly as a block 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  background-color: var(--bb-palette-white700,#000000);
  border-radius: 2px;
  color: var(--bb-palette-text,#000000);
  font-family: 'SF Mono','Segoe UI Mono','Roboto Mono',Menlo,Courier,monospace;
  padding: 0.1rem 0.2rem;
  overflow: auto;
}

.emotion-0 {
  padding: 1rem;
}

<pre
  class="emotion-0 emotion-1 emotion-2 bb-Box bb-Code"
>
  Hello world
</pre>
`;

exports[`props should render correctly with CSS props 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  background-color: var(--bb-palette-white700,#000000);
  border-radius: 2px;
  color: var(--bb-palette-text,#000000);
  font-family: 'SF Mono','Segoe UI Mono','Roboto Mono',Menlo,Courier,monospace;
  padding: 0.1rem 0.2rem;
  overflow: auto;
}

.emotion-2 {
  color: var(--bb-palette-primary,#000000) !important;
}

<code
  class="emotion-0 emotion-1 emotion-2 bb-Box bb-Code"
>
  Hello world
</code>
`;

exports[`props should render correctly with palette 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  background-color: var(--bb-palette-primaryTint,#000000);
  border-radius: 2px;
  color: var(--bb-palette-primary700,#000000);
  font-family: 'SF Mono','Segoe UI Mono','Roboto Mono',Menlo,Courier,monospace;
  padding: 0.1rem 0.2rem;
  overflow: auto;
}

<code
  class="emotion-0 emotion-1 emotion-2 bb-Box bb-Code"
>
  Hello world
</code>
`;

exports[`theming Code.base should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  background-color: var(--bb-palette-white700,#000000);
  border-radius: 2px;
  color: var(--bb-palette-text,#000000);
  font-family: 'SF Mono','Segoe UI Mono','Roboto Mono',Menlo,Courier,monospace;
  padding: 0.1rem 0.2rem;
  overflow: auto;
}

.emotion-0 {
  background-color: var(--bb-palette-red,#ff0000);
}

<code
  class="emotion-0 emotion-1 emotion-2 bb-Box bb-Code"
>
  hello world
</code>
`;
